vllm/core/scheduler.py

```diff
@@ -1078,44 +1079,85 @@ def schedule(self) -> Tuple[List[SequenceGroupMetadata], SchedulerOutputs]:
                 # a sequence is preempted, prefill includes previous generated
                 # output tokens.
-                if (token_chunk_size + seqs[0].data.get_num_computed_tokens() <
+                if (token_chunk_size + num_computed_tokens <
                         seqs[0].data.get_len()):
                     do_sample = False
 
             # It assumes the scheduled_seq_groups is ordered by
             # prefill < decoding.
-            is_prompt = seq_group.is_prefill()
-
-            seq_group_metadata.__init__(
-                request_id=seq_group.request_id,
-                is_prompt=is_prompt,
-                seq_data=seq_data,
-                sampling_params=seq_group.sampling_params,
-                block_tables=block_tables,
-                do_sample=do_sample,
-                pooling_params=seq_group.pooling_params,
-                token_chunk_size=token_chunk_size,
-                lora_request=seq_group.lora_request,
-                computed_block_nums=common_computed_block_nums,
-                encoder_seq_data=encoder_seq_data,
-                cross_block_table=cross_block_table,
-                # `multi_modal_data` will only be present for the 1st comm
-                # between engine and worker.
-                # the subsequent comms can still use delta, but
-                # `multi_modal_data` will be None.
-                multi_modal_data=seq_group.multi_modal_data
-                if scheduler_outputs.num_prefill_groups > 0 else None,
-                prompt_adapter_request=seq_group.prompt_adapter_request,
-            )
+            if is_first_prefill or not self.scheduler_config._send_delta_data:
+                # seq_group_metadata.__init__(
+                #     request_id=seq_group.request_id,
+                #     is_prompt=is_prompt,
+                #     seq_data=seq_data,
+                #     sampling_params=seq_group.sampling_params,
+                #     block_tables=block_tables,
+                #     do_sample=do_sample,
+                #     pooling_params=seq_group.pooling_params,
+                #     token_chunk_size=token_chunk_size,
+                #     lora_request=seq_group.lora_request,
+                #     computed_block_nums=common_computed_block_nums,
+                #     encoder_seq_data=encoder_seq_data,
+                #     cross_block_table=cross_block_table,
+                #     # `multi_modal_data` will only be present for the 1st comm
+                #     # between engine and worker.
+                #     # the subsequent comms can still use delta, but
+                #     # `multi_modal_data` will be None.
+                #     multi_modal_data=seq_group.multi_modal_data
+                #     if scheduler_outputs.num_prefill_groups > 0 else None,
+                #     prompt_adapter_request=seq_group.prompt_adapter_request,
+                # )
+                seq_group_metadata = SequenceGroupMetadata(
```

Comment:
NOTE: I removed _seq_group_metadata_cache.

- I ran this against the master, and I found there' no perf difference although I don't use the cache. I suspect it is because now `SequenceGroupMetadata` is a c object thanks to msgspec.Struct
- `__init__` doesn't work with msgspec.Struct

```
python3 benchmark_throughput.py --backend vllm --input-len 512 --output-len 256 --num-prompts 1000 --tensor-parallel 1
# this commit
Throughput: 20.72 requests/s, 15915.01 tokens/s
main
Throughput: 20.49 requests/s, 15735.44 tokens/s
```

Reply:
cc @alexm-neuralmagic